File: rtsp-streamer/onvif-service.js
const soap = require('soap');
const dgram = require('dgram');
const ip = require('ip');
const uuid = require('uuid');

class OnvifService {
  constructor(config) {
    this.config = config;
    this.rtspUrl = null;
    this.serverIp = config.serverIp || ip.address();
    this.onvifPort = config.onvifPort || 8555;
    this.deviceInfo = {
      manufacturer: 'YoLink',
      model: 'Dashboard-RTSP',
      firmwareVersion: '1.0.0',
      serialNumber: uuid.v4(),
      hardwareId: 'YOLINK-DASHBOARD-1'
    };
  }

  initialize(rtspUrl) {
    this.rtspUrl = rtspUrl;
    this.startSoapServer();
    this.startDiscovery();
    console.log(`ONVIF service initialized: onvif://${this.serverIp}:${this.onvifPort}`);
  }

  startSoapServer() {
    const service = {
      DeviceService: {
        DevicePort: {
          GetDeviceInformation: (args, callback) => {
            callback(null, {
              Manufacturer: this.deviceInfo.manufacturer,
              Model: this.deviceInfo.model,
              FirmwareVersion: this.deviceInfo.firmwareVersion,
              SerialNumber: this.deviceInfo.serialNumber,
              HardwareId: this.deviceInfo.hardwareId
            });
          },
          GetStreamUri: (args, callback) => {
            callback(null, {
              Uri: this.rtspUrl,
              InvalidAfterConnect: false,
              InvalidAfterReboot: false,
              Timeout: 'PT0S'
            });
          }
        }
      }
    };

    const wsdl = `
      <?xml version="1.0" encoding="UTF-8"?>
      <definitions xmlns="http://schemas.xmlsoap.org/wsdl/"
                   xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/"
                   targetNamespace="http://www.onvif.org/ver10/device/wsdl">
        <message name="GetDeviceInformationRequest"/>
        <message name="GetDeviceInformationResponse">
          <part name="parameters" element="Manufacturer Model FirmwareVersion SerialNumber HardwareId"/>
        </message>
        <message name="GetStreamUriRequest"/>
        <message name="GetStreamUriResponse">
          <part name="parameters" element="Uri"/>
        </message>
        <portType name="DevicePort">
          <operation name="GetDeviceInformation">
            <input message="tns:GetDeviceInformationRequest"/>
            <output message="tns:GetDeviceInformationResponse"/>
          </operation>
          <operation name="GetStreamUri">
            <input message="tns:GetStreamUriRequest"/>
            <output message="tns:GetStreamUriResponse"/>
          </operation>
        </portType>
        <binding name="DeviceBinding" type="tns:DevicePort">
          <soap:binding style="document" transport="http://schemas.xmlsoap.org/soap/http"/>
          <operation name="GetDeviceInformation">
            <soap:operation soapAction="http://www.onvif.org/GetDeviceInformation"/>
            <input><soap:body use="literal"/></input>
            <output><soap:body use="literal"/></output>
          </operation>
          <operation name="GetStreamUri">
            <soap:operation soapAction="http://www.onvif.org/GetStreamUri"/>
            <input><soap:body use="literal"/></input>
            <output><soap:body use="literal"/></output>
          </operation>
        </binding>
        <service name="DeviceService">
          <port name="DevicePort" binding="tns:DeviceBinding">
            <soap:address location="http://${this.serverIp}:${this.onvifPort}/onvif/device_service"/>
          </port>
        </service>
      </definitions>
    `;

    const soapServer = soap.listen(null, {
      path: '/onvif/device_service',
      port: this.onvifPort,
      services: service,
      xml: wsdl
    });

    soapServer.on('error', (err) => console.error('SOAP server error:', err));
    console.log(`SOAP server running on http://${this.serverIp}:${this.onvifPort}/onvif/device_service`);
  }

  startDiscovery() {
    const socket = dgram.createSocket('udp4');
    socket.bind(3702, () => socket.setMulticastTTL(128));

    socket.on('message', (msg, rinfo) => {
      if (msg.toString().includes('Probe')) {
        const response = `
          <s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:a="http://schemas.xmlsoap.org/ws/2004/08/addressing">
            <s:Header>
              <a:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/ProbeMatches</a:Action>
              <a:To>s:Sender</a:To>
            </s:Header>
            <s:Body>
              <d:ProbeMatches xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery">
                <d:ProbeMatch>
                  <a:EndpointReference><a:Address>urn:uuid:${this.deviceInfo.serialNumber}</a:Address></a:EndpointReference>
                  <d:Types>dn:NetworkVideoTransmitter</d:Types>
                  <d:Scopes>onvif://www.onvif.org/name/YoLinkDashboard</d:Scopes>
                  <d:XAddrs>http://${this.serverIp}:${this.onvifPort}/onvif/device_service</d:XAddrs>
                </d:ProbeMatch>
              </d:ProbeMatches>
            </s:Body>
          </s:Envelope>
        `;
        socket.send(response, 0, response.length, rinfo.port, rinfo.address);
      }
    });

    socket.on('listening', () => console.log('WS-Discovery listening on UDP 3702'));
    socket.on('error', (err) => console.error('Discovery error:', err));
    this.discoverySocket = socket;
  }

  stop() {
    if (this.discoverySocket) {
      this.discoverySocket.close();
      console.log('WS-Discovery stopped');
    }
    // SOAP server stop is handled by server shutdown in server.js
  }
}

module.exports = OnvifService;

File: onvif_service.py
"""
ONVIF service for device discovery and interaction.
"""
import uuid
import socket
import logging
import threading
from typing import Dict, Any

logger = logging.getLogger(__name__)


class OnvifService(threading.Thread):
    """
    ONVIF service for camera device discovery and interaction.
    Implements WS-Discovery for announcing the RTSP stream as an ONVIF camera.
    """

    def __init__(self, config: Dict[str, Any]):
        """
        Initialize the ONVIF service.

        Args:
            config: Application configuration
        """
        super().__init__()
        self.config = config
        self.server_ip = config.get("server_ip")
        self.onvif_port = config.get("onvif_port", 80)
        self.rtsp_port = config.get("rtsp_port", 554)
        self.stream_name = config.get("stream_name", "yolink-dashboard")

        # Generate unique device identifiers
        self.device_info = {
            "Manufacturer": "YoLink",
            "Model": "Dashboard-RTSP",
            "FirmwareVersion": "1.0.0",
            "SerialNumber": str(uuid.uuid4()),
            "HardwareId": "YOLINK-DASHBOARD-1"
        }

        # RTSP stream URL
        self.rtsp_url = f"rtsp://{self.server_ip}:{self.rtsp_port}/{self.stream_name}"

        # Thread settings
        self.daemon = True
        self.running = True

    def run(self) -> None:
        """
        Thread main function. Starts WS-Discovery service.
        """
        logger.info(f"Starting ONVIF service on port {self.onvif_port}")
        logger.info(f"ONVIF service initialized: onvif://{self.server_ip}:{self.onvif_port}")

        # Start WS-Discovery service in a separate thread
        discovery_thread = threading.Thread(target=self._ws_discovery, daemon=True)
        discovery_thread.start()

    def _ws_discovery(self) -> None:
        """
        Implement WS-Discovery for ONVIF device announcement.
        """
        # Create UDP socket for WS-Discovery
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        try:
            sock.bind(("", 3702))  # Standard WS-Discovery port
        except Exception as e:
            logger.error(f"WS-Discovery bind error: {e}")
            return

        # Enable broadcast
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        logger.info("WS-Discovery listening on UDP 3702")

        # Process incoming discovery requests
        while self.running:
            try:
                data, addr = sock.recvfrom(4096)

                # Look for Probe messages
                if b"Probe" in data:
                    logger.debug(f"Received WS-Discovery probe from {addr}")

                    # Send ProbeMatch response
                    response = self._generate_probe_match_response()
                    sock.sendto(response.encode(), addr)
                    logger.debug(f"Sent WS-Discovery response to {addr}")

            except Exception as e:
                if self.running:  # Only log if we're still supposed to be running
                    logger.error(f"WS-Discovery error: {e}")

    def _generate_probe_match_response(self) -> str:
        """
        Generate a WS-Discovery ProbeMatch response.

        Returns:
            str: SOAP XML response
        """
        return f"""
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:a="http://schemas.xmlsoap.org/ws/2004/08/addressing">
  <s:Header>
    <a:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/ProbeMatches</a:Action>
    <a:To>s:Sender</a:To>
  </s:Header>
  <s:Body>
    <d:ProbeMatches xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery">
      <d:ProbeMatch>
        <a:EndpointReference><a:Address>urn:uuid:{self.device_info['SerialNumber']}</a:Address></a:EndpointReference>
        <d:Types>dn:NetworkVideoTransmitter</d:Types>
        <d:Scopes>onvif://www.onvif.org/name/YoLinkDashboard</d:Scopes>
        <d:XAddrs>http://{self.server_ip}:{self.onvif_port}/onvif/device_service</d:XAddrs>
      </d:ProbeMatch>
    </d:ProbeMatches>
  </s:Body>
</s:Envelope>
"""

    def stop(self) -> None:
        """
        Stop the ONVIF service.
        """
        logger.info("Stopping ONVIF service")
        self.running = False